### PR TITLE
Abstract data types and utilities

### PR DESCRIPTION
**Pull request by breuleux**
In preparation for the inference algorithm, here are the data structures for the type system as well as the functions to manipulate them.

### CODE REVIEW

File: myia/abstract/data.py [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+        """Initialize an AbstractValue."""
+        super().__init__()
+        if not isinstance(tracks, Tracks):
+            tracks = Tracks(**tracks)
+        self.tracks = tracks
+
+    @property
+    def t(self):
+        """Returns the AbstractValue's tracks."""
+        return self.tracks
```
Comment (abergeron) [inline myia/abstract/data.py]: I get that this is a shortcut, but is it really necessary?

File: myia/abstract/data.py [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+                type=f"*{self.tracks.interface.__name__}",
+            )
+
+
+class AbstractUnion(AbstractValue):
+    """Base class for abstract values that are sum types."""
+
+    def __init__(self, options, tracks):
+        super().__init__(tracks)
+        self.options = list(options)  # Possibilities(options)
```
Comment (abergeron) [inline myia/abstract/data.py]: That comment seems like leftovers.

**Review by abergeron — APPROVED** [review]

File: myia/abstract/data.py (hunk 1 above)
Comment (breuleux) [reply myia/abstract/data.py]: I suppose not

File: myia/abstract/data.py (hunk 2 above)
Comment (breuleux) [reply myia/abstract/data.py]: Ah, yeah

**Pull request merged**